### PR TITLE
OAK 0.4.2

### PR DESCRIPTION
**Pull request by jhwillett**
PR for `oak` (which is already open and published at 0.4.1).

A grab-bag.

Reviewers: please take time to read `DESIGN.rb` and `ENCRYPTION.rb` with the "VIEW" button.  Please keep an eye out for information which maybe I shouldn't be pushing into an open project.

At present there are things like email addresses of PW personnel and excerpts from other company's public documents: Those things were uncontroversial in our private Google Docs space.  Are they problematic in this public venue?

### CODE REVIEW

File: bin/oak.rb [hunk 1]
```diff
@@ -103,143 +103,141 @@
 
 =end
 
-if __FILE__ == $0
```
Comment (jhwillett) [inline bin/oak.rb]: The only change here was to get rid of the `if __FILE__ == $0` condition.

When `oak` has been installed as a gem, `bundler` creates a little forwarding script which gets put in the PATH:
```
hich oak
/Users/jhw/.rvm/gems/ruby-2.2.7/bin/oak
$ tail -3 `which oak`
gem "oak", version
load Gem.bin_path("oak", "oak", version)
end
```
When _this_ script is invoked from _that_ script, `$0` is something like `/Users/jhw/.rvm/gems/ruby-2.2.7/bin/oak` but `__FILE__` is something like `/Users/jhw/.rvm/gems/ruby-2.2.7/bin/oak.rb`.  They don't match, and so the script does nothing!

:(

Comment (jhwillett) [reply bin/oak.rb]: This file's change is easier to view in GitHub with `?w=1`.

File: DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+JSON or YAML, but slightly more Ruby-esque, and supports fewer types
+than Marshal.  OAK also has features similar to OpenPGP
+([https://tools.ietf.org/html/rfc4880](https://tools.ietf.org/html/rfc4880))
+(though not, so far, encryption).
+
+The main value proposition for OAK is operational flexibility.  OAK
+leds you defer choices between compression, checksumming, and 7-bit
+cleanliness algorithms until after a system is live and under load.
+
+As of 2017-09-13, OAK is used by ALI for volatile caches in Redis, and
```
Comment (marshall-prosperworks) [inline DESIGN.md]: Codename `ALI` for "The main part of the Copper CRM" might not be meaningful

File: DESIGN.md [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+for durable Correspondence bodies in S3.
+
+Author:
+- [jhw@prosperworks.com](mailto:jhw@prosperworks.com)
+
+Advisors:
+- [marshall@prosperworks.com](mailto:marshall@prosperworks.com)
```
Comment (marshall-prosperworks) [inline DESIGN.md]: @copper.com email addresses maybe?

File: DESIGN.md [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+- [gerald@prosperworks.com](mailto:gerald@prosperworks.com)
```
Comment (marshall-prosperworks) [inline DESIGN.md]: Might as well not list Gerald here, unless he is agreeing to continue to support Oak, in which case we might want to update his contact info

File: DESIGN.md [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+    * Introduced RedisCache.  Later PRs use RedisCache for
+        * SummaryAccessor (OAK)
+        * RussianDoll caches (Marshal+OAK)
+        * S3 cache (OAK)
+        * COMPANY_ID_CACHE (JSON)
+    * Commits to OAK for volatile use cases.
+* [https://github.com/ProsperWorks/ALI/pull/1757](https://github.com/ProsperWorks/ALI/pull/1757) **oak-woe-2016-06-30**
+    * Merged in major_2016_07_hotpocket.
+    * Fixed a regexp which had broken some Float parsing.
+* [https://github.com/ProsperWorks/ALI/pull/1724](https://github.com/ProsperWorks/ALI/pull/1724) **correspondences-in-s3-fixes**
```
Comment (marshall-prosperworks) [inline DESIGN.md]: A non-Prosperworks employee will not have access to the `ALI` project in Github, certainly

File: DESIGN.md [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+
+>> str = 'x' ; OAK.encode([str,str],format: :none)
+=> "oak_3CNN_2865617390_13_F2A2_1_1SU1_x_ok"
+```
+
+The JSON format does not support `Infinity`, `-Infinity`, or `NaN` -
+though Ruby's JSON encoder transcodes thes via a nonstandard
+extension.
+
+YAML handles `Infinity`, `-Infinity`, and `NaN`.  YAML also also DAGs
```
Comment (marshall-prosperworks) [inline DESIGN.md]: `also also` looks like duplicate `also` and missing `handles`

File: DESIGN.md [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+
+>> JSON.dump(a)
+SystemStackError: stack level too deep
+
+>> OAK.encode(a,format: :none)
+=> "oak_3CNN_3573295141_24_F4A2_1_2SU1_aA2_3_0SU1_b_ok"
+```
+The essence of serializing non-tree objects is capturing identity.
+
+Does this matter in ALI?  Honestly, I don't know.  Cycles and DAGs are
```
Comment (marshall-prosperworks) [inline DESIGN.md]: Codename `ALI`

File: DESIGN.md [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+      source."
+    * Marshal can have problems if a user-defined class changes between
+      encoding time and decoding time.
+    * I wanted OAK to refuse to encode objects whose structure it
+      could not guarantee to recover with perfect fidelity.
+* I have ambitions for language portability with in OAK.
+    * Specious: porting a subset of Marshal would be no harder than
+      porting OAK.
+
+To be fair, we use Marshal anyhow, wrapped in OAK, in the Russian Doll
```
Comment (marshall-prosperworks) [inline DESIGN.md]: Users might not have any concept for Russian Doll or Summary caches

File: ENCRYPTION.md [hunk 9]
```diff
@@ -0,0 +1,898 @@
...
+message.**_
+
+Therefore _**encrypted OAK will be nondeterministic in the plaintext**_.
+This is a bummer but I see no way to avoid it without compromising
+security.
+
+Also, today I see no point in supporting anything other than AES.  All
+of AES-128, AES-192, and AES-256 are probably adequate for our needs,
+but if we support just AES-256 then we don't have to answer any thorny
+questions.  There too much securit downside in letting the caller pick
```
Comment (marshall-prosperworks) [inline ENCRYPTION.md]: `securit` -> `security`

**Pull request merged**